### PR TITLE
Loki: LogQL v2 support

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:

This PR is pre-work for support of [LogQL v2](https://github.com/grafana/loki/blob/a87cf44829841958a3dd78030266ac93eb7269c3/docs/design-documents/2020-03-logql-fields.md). It implements following:

- Adds new RANGE_VEC_FUNCTIONS (avg_over_time, min_over_time, max_over_time, sum_over_time)
- Adds Loki PARSERS + creates regex for parsers parsing. In primsm, adds `keyword` className for highlighting purposes. Add Parsers to suggestions.
- Updates regex pattern for context-labels, so the ending `}` is optional. This also fixes **Autocomplete does not work on incomplete input** bug for Loki https://github.com/grafana/grafana/issues/20445. 
- Updates regex pattern for operators. On master, it is currently highlighting various operators, but not the loki specific filter operators weren't correctly matched.  
- Adds test coverage

image

**Next PR**:
-  Technically you could suggest identifier after `|` by sending a query and looking at all labels available - we can implement this is next PR. 
- Errors when using parsers will include objects with log lines. Make sure to display them in errors.

### CODE REVIEW

File: public/app/plugins/datasource/loki/syntax.ts [hunk 1]
```diff
@@ -49,7 +49,67 @@ const AGGREGATION_OPERATORS: CompletionItem[] = [
...
+    documentation: 'Extracting labels from the log line using regexp parser. Only available in Loki 2.0+.',
+  },
+  {
+    label: 'logfmt',
+    insertText: 'logfmt',
+    documentation: 'Extracting labels from the log line using logfmt parser. Only available in Loki 2.0+.',
+  },
+];
+
+export const PIPE_OPERATORS: CompletionItem[] = [
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/syntax.ts:70]: @cyriltovena Are there any other pipe operators that we should suggest after user writes `|`? Also so we have any documentation/hints for these?

File: public/app/plugins/datasource/loki/syntax.ts [hunk 2]
```diff
@@ -49,7 +49,67 @@ const AGGREGATION_OPERATORS: CompletionItem[] = [
...
+export const PIPE_PARSERS: CompletionItem[] = [
+  {
+    label: 'json',
+    insertText: 'json',
+    documentation: 'Extracting labels from the log line using json parser. Only available in Loki 2.0+.',
+  },
+  {
+    label: 'regexp',
+    insertText: 'regexp',
+    documentation: 'Extracting labels from the log line using regexp parser. Only available in Loki 2.0+.',
```
Comment (cyriltovena) [inline public/app/plugins/datasource/loki/syntax.ts:61]: Could we automatically move the cursor into the regex string. e.g `| regexp "here"` ?

File: public/app/plugins/datasource/loki/syntax.ts (hunk 1 above)
Comment (cyriltovena) [reply public/app/plugins/datasource/loki/syntax.ts:70]: label_filter is basically `| duration > 250ms` not sure what you can suggest. Technically you could suggest identifier by sending a query and looking at all labels available but that sounds complex for a first round ?

Basically `| label_filter` doesn't exists.

The last operator we have is `| unwrap identifier` which allows you to take a labels and use the value as sample data for metric aggregations.

File: public/app/plugins/datasource/loki/syntax.ts [hunk 3]
```diff
@@ -49,7 +49,67 @@ const AGGREGATION_OPERATORS: CompletionItem[] = [
...
+    documentation: 'The minimum of all values in the specified interval.',
+  },
+  {
+    insertText: 'max_over_time',
+    label: 'max_over_time',
+    detail: 'max_over_time(range-vector)',
+    documentation: 'The maximum of all values in the specified interval.',
+  },
+  {
+    insertText: 'sum_over_time',
```
Comment (cyriltovena) [inline public/app/plugins/datasource/loki/syntax.ts:108]: We also have those:

`stdvar_over_time(range vec)` , `stddev_over_time(range vec)` and `quantile_over_time(value between 0 and 1,range vec)`

In the doc I was saying we won't do it but I just did them anyway as I think they are killer feature.

Comment (cyriltovena) [reply public/app/plugins/datasource/loki/syntax.ts:108]: Full list here: https://github.com/grafana/loki/compare/master...cyriltovena:logql-parser#diff-8c1350fd75c7775d87e4408faa8c5fc2R293

File: public/app/plugins/datasource/loki/syntax.ts (hunk 2 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/syntax.ts:61]: Yes, good idea. Added in https://github.com/grafana/grafana/pull/27884/commits/796afda6df34f9ef23a23bef5c6b67d582ee8d90

File: public/app/plugins/datasource/loki/syntax.ts (hunk 1 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/syntax.ts:71]: I have removed `label_filter` and added `unwrap identifier`.

File: public/app/plugins/datasource/loki/syntax.ts (hunk 3 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/syntax.ts:110]: I have added those. One question, should we add also `BYTES_OVER_TIME` and `BYTES_RATE`?

Comment (cyriltovena) [reply public/app/plugins/datasource/loki/syntax.ts:110]: yes we should.

BYTES_OVER_TIME => gives you the bytes count of logs over time
BYTES_RATE => gives you the bytes per second of logs over time

File: public/app/plugins/datasource/loki/syntax.ts [hunk 4]
```diff
@@ -69,9 +70,10 @@ export const PIPE_PARSERS: CompletionItem[] = [
 
 export const PIPE_OPERATORS: CompletionItem[] = [
   {
-    label: 'label_filter',
-    insertText: 'label_filter',
-    documentation: 'Only available in Loki 2.0+.',
+    label: 'unwrap identifier',
+    insertText: 'unwrap identifier',
```
Comment (cyriltovena) [inline public/app/plugins/datasource/loki/syntax.ts:74]: sorry I wasn't clear @ivanahuckova identifier means the user will have to choose a label identifier.

e.g: `| unwrap latency`

**Comment by cyriltovena** [comment]
The last bit for supporting logql v2 will be error handling.

Currently this is not implemented in dev. but the idea is that some log request might come out as 400 but with still a set of stream as response.

File: public/app/plugins/datasource/loki/syntax.ts (hunk 3 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/syntax.ts:111]: Thanks! Not sure why they weren't there before, but I have added them as well. 🙂 Also regarding this: 

>In the doc I was saying we won't do it but I just did them anyway as I think they are killer feature.

🔥🔥🔥🔥🔥

**Review by zoltanbedi — APPROVED** [review]
lgtm 🔱

**Pull request merged**